opt_einsum/paths.py

```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+    two = k1 & k2
+    one = either - two
+    k12 = (either & output) | (two & dim_ref_counts[3]) | (one & dim_ref_counts[2])
+    footprint12 = helpers.compute_size_by_dict(k12, sizes)
+    cost = footprint12 - footprints[k1] - footprints[k2]
+    id1 = remaining[k1]
+    id2 = remaining[k2]
+    cost = cost, min(id1, id2), max(id1, id2)  # break ties to ensure determinism
+    return cost, k1, k2, k12
+
+
+def _push_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2s, queue):
+    if not k2s:
+        return
+    candidate = min(_get_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2)
+                    for k2 in k2s)
+    heapq.heappush(queue, candidate)
+
+
+def _update_ref_counts(dim_to_keys, dim_ref_counts, dims):
+    for dim in dims:
+        count = len(dim_to_keys[dim])
+        if count <= 1:
+            dim_ref_counts[2].discard(dim)
+            dim_ref_counts[3].discard(dim)
+        elif count == 2:
+            dim_ref_counts[2].add(dim)
+            dim_ref_counts[3].discard(dim)
+        else:
+            dim_ref_counts[2].add(dim)
+            dim_ref_counts[3].add(dim)
+
+
+def _ssa_optimize(inputs, output, sizes):
+    """
+    This has an interface similar to :func:`cheap` but produces a path with
+    static single assignment ids rather than recycled linear ids.
+    SSA ids are cheaper to work with and easier to reason about.
+    """
+    # A dim common to all terms might as well be an output dim.
+    inputs = list(map(frozenset, inputs))
+    output = frozenset(output) | frozenset.intersection(*inputs)
+
+    # Deduplicate shapes by eagerly computing Hadamard products.
+    remaining = {}  # key -> ssa_id
+    ssa_ids = itertools.count(len(inputs))
+    ssa_path = []
+    for ssa_id, key in enumerate(inputs):
+        if key in remaining:
+            ssa_path.append((remaining[key], ssa_id))
+            remaining[key] = next(ssa_ids)
+        else:
+            remaining[key] = ssa_id
+
+    # Compute footprints of each tensor.
+    footprints = {key: helpers.compute_size_by_dict(key, sizes)
+                  for key in remaining}
+
+    # Keep track of possible contraction dims.
+    dim_to_keys = defaultdict(set)
+    for key in remaining:
+        for dim in key - output:
+            dim_to_keys[dim].add(key)
+
+    # Keep track of the number of tensors using each dim; when the dim is no longer
+    # used it can be contracted. Since we specialize to binary ops, we only care about
+    # ref counts of >=2 or >=3.
+    dim_ref_counts = {
+        count: set(dim for dim, keys in dim_to_keys.items() if len(keys) >= count) - output
+        for count in [2, 3]}
+
+    # Find initial candidate contractions.
+    queue = []
+    for dim, keys in dim_to_keys.items():
+        keys = list(keys)
+        for i, k1 in enumerate(keys):
+            k2s = keys[:i]
+            _push_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2s, queue)
```

Comment:
Still wrapping my head around this part a bit. Could you not just iterate over each ``dim`` that appears more than twice (``dim_ref_counts[2]``)?

I think I see that that you only ever need to add the minimum cost from ``k1`` between ``k2s``, is this the reason for not just pushing all the pairwise candidates?

Reply:
Originally I pushed all the candidates, but ended up with something like 30000 candidates in my test case. This ended up slowing down the O(n log(n)) heap logic. Switching to eager O(n) minimization gave me about a 10x speedup on my large test case.